mirgecom/euler.py

```diff
@@ -20,16 +22,307 @@
 THE SOFTWARE.
 """
 
-# import numpy as np
-# import numpy.linalg as la  # noqa
-# from pytools.obj_array import (
-#         join_fields, make_obj_array,
-#         with_object_array_or_scalar)
-# from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
-# from grudge.eager import with_queue
-# from grudge.symbolic.primitives import TracePair
+import numpy as np
+import numpy.linalg as la  # noqa
+from pytools.obj_array import (
+    flat_obj_array,
+    make_obj_array,
+    with_object_array_or_scalar,
+)
+import pyopencl.clmath as clmath
+import pyopencl.array as clarray
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+from mirgecom.eos import IdealSingleGas
+
+from grudge.eager import with_queue
+from grudge.symbolic.primitives import TracePair
+from dataclasses import dataclass
+
+
+r"""
+This module is designed provide functions and utilities
+useful for solving the Euler flow equations.
+
+The Euler flow equations are:
+
+.. :math::
+
+    \partial_t \mathbf{Q} = -\nabla\cdot{\mathbf{F}} +
+    (\mathbf{F}\cdot\hat{n})_\partial_{\Omega} + \mathbf{S}
+
+where:
+    state :math:`\mathbf{Q} = [\rho, \rho{E}, \rho\vec{V} ]`
+    flux :math:`\mathbf{F} = [\rho\vec{V},(\rho{E} + p)\vec{V},
+                (\rho(\vec{V}\otimes\vec{V}) + p*\mathbf{I})]`,
+    domain boundary :math:`\partial_{\Omega}`,
+    sources :math:`mathbf{S} =
+                   [{(\partial_t{\rho})}_s, {(\partial_t{\rho{E}})}_s,
+                    {(\partial_t{\rho\vec{V}})}_s]`
+
+"""
+
+# from grudge.dt_finding import (
+#    dt_geometric_factor,
+#    dt_non_geometric_factor,
+# )
+
+__doc__ = """
+.. autofunction:: inviscid_operator
+.. autofunction:: number_of_scalars
+.. autofunction:: split_conserved
+.. autofunction:: split_species
+.. autofunction:: split_fields
+.. autofunction:: get_inviscid_timestep
+"""
+
+
+#
+# Euler flow eqns:
+# d_t(q) + nabla .dot. f = 0 (no sources atm)
+# state vector q: [rho rhoE rhoV]
+# flux tensor f: [rhoV (rhoE + p)V (rhoV.x.V + p*I)]
+#
+
+
+@dataclass
+class ConservedVars:
+    r"""
+    Class to resolve the canonical conserved quantities,
+    (mass, energy, momentum) per unit volume =
+    :math:`(\rho,\rhoE,\rho\vec{V})` from an agglomerated
+    object array.
+    """
+    mass: np.ndarray
+    """mass per unit volume"""
+    energy: np.ndarray
+    """energy per unit volume"""
+    momentum: np.ndarray
+    """momentum vector per unit volume"""
```

Comment:
I don't think sphinx picks up on docstrings the way there written here. We should figure out what the right way is. One that I know works is to have

```rst
.. attribute:: mass

    Mass per Unit Volume.
```
in the class docstring.